source/vsignal/signal.d

```diff
@@ -0,0 +1,112 @@
+module vsignal.signal;
+
+import vsignal.sink;
+import vsignal.slot;
+
+struct Signal(F)
+	if (is(F : RT delegate(Args), RT, Args...))
+{
+	/**
+	Publish a signal.
+
+	Params:
+		args = the Signal's function parameters to call each listener with.
+	*/
+	void emit(Slot!F.Parameters args)
+	{
+		import core.lifetime : forward;
+
+		foreach (call; calls)
+			cast(void) call(forward!args);
+	}
+
+	bool empty() const @property
+	{
+		return !calls.length;
+	}
+
+	size_t length() const @property
+	{
+		return calls.length;
+	}
+
+	Sink!F sink() @property
+	{
+		return Sink!F(&this);
+	}
+
+package:
+	Slot!F[] calls;
+}
+
+///
+@safe pure nothrow unittest
+{
+	Signal!(void delegate(int) @safe pure nothrow) signal;
+	Signal!(void delegate(ref int) @safe pure nothrow) signalRef;
+
+	static struct Listener
+	{
+		int i;
+		void opCall(const ref int)
+		{
+			i++;
+		}
+	}
+
+	Listener listener;
+
+	signal.sink.connect!(Listener.opCall)(listener);
+	signalRef.sink.connect!(Listener.opCall)(listener);
+
+	int var;
+
+	signal.emit(var);
+	signalRef.emit(var);
+
+	assert(listener.i == 2);
+}
+
+///
+@safe nothrow unittest
+{
+	Signal!(void delegate(int) @safe nothrow @nogc) signal;
+	int var = 35;
+
+	signal.sink.connect!((ref i, const int x) @trusted {
+		assert(&i is &var);
+		i += x;
+	})(var);
+
+	signal.emit(3);
+
+	assert(var == 38);
+}
+
+unittest
+{
+	Signal!(void delegate(int)) sig;
+
+	struct Listener
+	{
+		int i;
+		void opCall(int var) { i = var; }
+	}
+
+	Listener listener;
+
+	Slot!(void delegate(int)) slotA;
+	Slot!(void delegate(int)) slotB;
+
+	slotA.connect!(Listener.opCall)(listener);
+	slotB.connect!((ref i) => listener.i++ );
+
+	sig.calls = [
+		slotA,
+		slotB
+	];
```

Comment:
Just to confirm; is the `Signal` ensuring that calls are made in the order these were stored? Can't find anywhere on the docs whether this is guaranteed, but I recall having issues with order on arrays.

Reply:
I haven't documented `Signal` yet, but you cannot rely on the order. They are published in the same order they are stored, but their order is not guaranteed to stay the same throughout the usage of the Signal.

Reply:
Then this unittest might fail, since it relies on execution order for it's final result.

Reply:
No, as I said they are published in the same order they are stored. So `slotA` and then `slotB`. But the order is not guaranteed to stay the same throughout the usage of the Signal.

This happens only when removing listeners. The algorithm I used prefers speed to order. This is not a problem because the user should not rely on the order of listeners for implementations when working with a Signal/Slots structure.

But the test will never fail.